arch/xtensa/src/esp32s2/esp32s2_i2s.c

```diff
@@ -0,0 +1,1530 @@
+  uint32_t dout_outsig;       /* TX channel Data Output signal index */
+  uint32_t mclk_out_sig;      /* Master clock output index */
+
+  bool bit_shift;             /* Set to enable bit shift in Philips mode */
+  bool mono_en;               /* Set to enable mono mode on slot */
+
+  /* WS signal width (the number of bclk ticks that ws signal is high) */
+
+  uint32_t ws_width;
+
+  /* WS signal polarity, set true to enable high lever first */
+
+  bool ws_pol;
+};
+
+struct esp32s2_buffer_s
+{
+  struct esp32s2_buffer_s *flink; /* Supports a singly linked list */
+
+  /* The associated DMA outlink */
+
+  struct esp32s2_dmadesc_s dma_outlink[I2S_DMADESC_NUM];
+
+  i2s_callback_t callback;        /* DMA completion callback */
+  uint32_t timeout;               /* Timeout value of the DMA transfers */
+  void *arg;                      /* Callback's argument */
+  struct ap_buffer_s *apb;        /* The audio buffer */
+  int result;                     /* The result of the transfer */
+};
+
+/* This structure describes the state of one receiver or transmitter
+ * transport.
+ */
+
+struct esp32s2_transport_s
+{
+  sq_queue_t pend;              /* A queue of pending transfers */
+  sq_queue_t act;               /* A queue of active transfers */
+  sq_queue_t done;              /* A queue of completed transfers */
+  struct work_s work;           /* Supports worker thread operations */
+};
+
+/* The state of the one I2S peripheral */
+
+struct esp32s2_i2s_s
+{
+  struct i2s_dev_s  dev;        /* Externally visible I2S interface */
+  sem_t             exclsem;    /* Assures mutually exclusive access */
+  uint32_t          rate;       /* I2S actual configured sample-rate */
+  uint32_t          data_width; /* I2S actual configured data_width */
+  int               cpuint;     /* I2S interrupt ID */
+  uint8_t           cpu;        /* CPU ID */
+
+  /* Port configuration */
+
+  const struct esp32s2_i2s_config_s *config;
+
+#ifdef I2S_HAVE_TX
+  struct esp32s2_transport_s tx;        /* TX transport state */
+
+  /* Stuff var to fill DMA buffer if not word-aligned */
+
+  uint32_t stuff;
+#endif /* I2S_HAVE_TX */
+
+  /* Pre-allocated pool of buffer containers */
+
+  sem_t bufsem;                         /* Buffer wait semaphore */
+  struct esp32s2_buffer_s *bf_freelist; /* A list a free buffer containers */
+  struct esp32s2_buffer_s containers[CONFIG_ESP32S2_I2S_MAXINFLIGHT];
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+/* Register helpers */
+
+#ifdef CONFIG_ESP32S2_I2S_DUMPBUFFERS
+#  define       i2s_dump_buffer(m,b,s) lib_dumpbuffer(m,b,s)
+#else
+#  define       i2s_dump_buffer(m,b,s)
+#endif
+
+/* Semaphore helpers */
+
+static int      i2s_exclsem_take(struct esp32s2_i2s_s *priv);
+#define         i2s_exclsem_give(priv) nxsem_post(&priv->exclsem)
```

Comment:
I think we should use `nxmutex`
cc: @xiaoxiang781216

Reply:
you're right!